lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskStatus.cs

```diff
@@ -98,28 +98,48 @@ public void SetException(Exception e)
         {
             lock (_heartBeatManager)
             {
-                try
+                if (_lastException.IsPresent())
                 {
-                    if (!_lastException.IsPresent())
-                    {
-                        _lastException = Optional<Exception>.Of(e);
-                    }
+                    return;
+                }
 
-                    State = TaskState.Failed;
+                _lastException = Optional<Exception>.Of(e);
+                State = TaskState.Failed;
+
+                try
+                {
                     _taskLifeCycle.Stop();
                 }
-                finally
+                catch (Exception stopException)
                 {
-                    Heartbeat();
+                    Utilities.Diagnostics.Exceptions.Caught(
```

Comment:
Task.Close() is also called when ActiveContext.Dispose() is called. In most cases, the task should have disposed before it, therefore we skip the task.Close(). In case the task is still running, when task.Close is called, if the client doesn't implement the task close handler, the default task close handler will be called. Currently, it throws not implement exception. 
Even if the client has provided task close handler, there is still a possibility that the client code doesn't close the task properly, and the task thread is still running. Shall we enforce to stop the task thread when default stop handler is called?